core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala

```diff
@@ -422,7 +484,77 @@ object SubtractFeeFromOutputsFinalizer {
       .dropRight(1)
       .:+((newLastOutput, lastOutputIndex))
 
-    (newOutputs ++ unchangedOutputs).sortBy(_._2).map(_._1).toVector
+    (newOutputs ++ unchangedOutputs).sortBy(_._2).map(_._1)
+  }
+}
+
+case class DualFundingInput(
+    scriptSignature: ScriptSignature,
+    maxWitnessLen: Int)
+
+/** Finalizes a dual-funded transaction given the DualFundingInputs
+  * from both parties, their change spks and the non-change spk.
+  *
+  * This includes adding the future fee of spending transactions
+  * to the non-change output as well as subtracting relevant fees
+  * from the change outputs. This finalizer filters dust outputs.
+  */
+case class DualFundingTxFinalizer(
+    offerInputs: Vector[DualFundingInput],
+    offerPayoutSPK: ScriptPubKey,
+    offerChangeSPK: ScriptPubKey,
+    acceptInputs: Vector[DualFundingInput],
+    acceptPayoutSPK: ScriptPubKey,
+    acceptChangeSPK: ScriptPubKey,
+    feeRate: SatoshisPerVirtualByte,
+    fundingSPK: ScriptPubKey)
+    extends RawTxFinalizer {
+
+  private def computeFees(
+      inputs: Vector[DualFundingInput],
+      payoutSPK: ScriptPubKey,
+      changeSPK: ScriptPubKey): (CurrencyUnit, CurrencyUnit) = {
+    val futureFeeWeight = 249 + 4 * payoutSPK.asmBytes.length
+    val futureFeeVBytes = Math.ceil(futureFeeWeight / 4.0).toLong
+    val futureFee = feeRate * futureFeeVBytes
+
+    val inputWeight =
+      inputs.foldLeft(0L) {
+        case (weight, DualFundingInput(scriptSignature, maxWitnessLen)) =>
+          weight + 164 + 4 * scriptSignature.asmBytes.length + maxWitnessLen.toInt
+      }
+    val outputWeight = 36 + 4 * changeSPK.asmBytes.length
+    val weight = 107 + outputWeight + inputWeight
+    val vbytes = Math.ceil(weight / 4.0).toLong
+    val fundingFee = feeRate * vbytes
+
+    (futureFee, fundingFee)
+  }
+
+  lazy val (offerFutureFee, offerFundingFee) =
+    computeFees(offerInputs, offerPayoutSPK, offerChangeSPK)
+
+  lazy val (acceptFutureFee, acceptFundingFee) =
+    computeFees(acceptInputs, acceptPayoutSPK, acceptChangeSPK)
+
+  override def buildTx(txBuilderResult: RawTxBuilderResult)(implicit
+      ec: ExecutionContext): Future[Transaction] = {
+    val addOfferFutureFee =
+      AddFutureFeeFinalizer(fundingSPK, offerFutureFee, Vector(offerChangeSPK))
+    val addAcceptFutureFee = AddFutureFeeFinalizer(fundingSPK,
+                                                   acceptFutureFee,
+                                                   Vector(acceptChangeSPK))
+    val subtractOfferFundingFee =
+      SubtractFromOutputFinalizer(offerChangeSPK, offerFundingFee)
+    val subtractAcceptFundingFee =
+      SubtractFromOutputFinalizer(acceptChangeSPK, acceptFundingFee)
+
+    addOfferFutureFee
+      .andThen(addAcceptFutureFee)
```

Comment:
This is nice! I feel like `.andThen()` could be renamed to `.map()`. Nice job!

Reply:
Thanks :) I'm not a fan of map here as this the thing on which the method is being called is not the tx, it is the finalizer (making this a composition of functions rather than a mapping on a wrapper of some kind)